### PR TITLE
Feat: propose AddonDefinition

### PR DESCRIPTION
**Pull request by hongchaodeng**
ref: https://github.com/oam-dev/catalog/pull/115

### CODE REVIEW

File: design/vela-core/addon.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+metadata:
+  name: fluxcd
+  namespace: vela-system
+spec:
+  version: 1.2.3
+  description: Extended workload to do continuous and progressive delivery
+
+  # contact:
+  #   author: xxx 
+  #   email: xxx@xxx.com
```
Comment (Somefive) [inline design/vela-core/addon.md:37]: Is `contact` an optional field?

File: design/vela-core/addon.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    url: https://kubevela.io/docs/end-user/components/kustomize
+
+  dependencies:
+  - namespace: addon_namespace
+    name: addon_name
+
+  # The addon installation process and parameterization are defined in CUE.
+  # This would provide flexibility and consistency for the addon management system within KubeVela.
+  template: |
+    apply: op.#Apply {
```
Comment (Somefive) [inline design/vela-core/addon.md:63]: import ("vela/op") ?

File: design/vela-core/addon.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+```yaml
+kind: AddonDefinition
+metadata:
+  name: fluxcd
+  namespace: vela-system
+spec:
+  version: 1.2.3
+  description: Extended workload to do continuous and progressive delivery
+
+  # contact:
```
Comment (wonderflow) [inline design/vela-core/addon.md:35]: provider?

File: design/vela-core/addon.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  dependencies:
+  - namespace: addon_namespace
+    name: addon_name
+
+  # The addon installation process and parameterization are defined in CUE.
+  # This would provide flexibility and consistency for the addon management system within KubeVela.
+  template: |
+    apply: op.#Apply {
+      value: {
+        kind: Application
```
Comment (wonderflow) [inline design/vela-core/addon.md:65]: seems we don't need to use Application if we render it as template here. Is there anyway to re-use the Application Definition to do so?

**Comment by chivalryq** [comment]
If we use cue, it will be like
```cue
annotations: {
	type:        "addon"
	name:        "terraform"
	description: "Terraform Controller is a Kubernetes Controller for Terraform."
	definitions: [
		{
			name:        "helm"
			type:        "component"
			description: "helps to deploy a helm chart from everywhere: git repo, helm repo, or S3 compatible bucket."
		},
	]
	dependencies: [
		...,
	]
}
template: {
	application: {
		metadata: {
			name:      "terraform"
			namespace: "vela-system"
		}
		spec: {
			components: {
				xxx: yyy
			}
			workflow: {
				...
			}
		}
	}
	parameter: {
		appliedNamespace: *"vela-system" | string
	}
}

```

**Comment by Somefive** [comment]
> If we use cue, it will be like
> 
> ```cue
> annotations: {
> 	type:        "addon"
> 	name:        "terraform"
> 	description: "Terraform Controller is a Kubernetes Controller for Terraform."
> 	definitions: [
> 		{
> 			name:        "helm"
> 			type:        "component"
> 			description: "helps to deploy a helm chart from everywhere: git repo, helm repo, or S3 compatible bucket."
> 		},
> 	]
> 	dependencies: [
> 		...,
> 	]
> }
> template: {
> 	application: {
> 		metadata: {
> 			name:      "terraform"
> 			namespace: "vela-system"
> 		}
> 		spec: {
> 			components: {
> 				xxx: yyy
> 			}
> 			workflow: {
> 				...
> 			}
> 		}
> 	}
> 	parameter: {
> 		appliedNamespace: *"vela-system" | string
> 	}
> }
> ```

`application` should be wrapped in `output`?

File: design/vela-core/addon.md (hunk 1 above)
Comment (hongchaodeng) [reply design/vela-core/addon.md:37]: yes

File: design/vela-core/addon.md (hunk 3 above)
Comment (hongchaodeng) [reply design/vela-core/addon.md:35]: This is not provider. Just the contact info of the author.

File: design/vela-core/addon.md (hunk 4 above)
Comment (hongchaodeng) [reply design/vela-core/addon.md:65]: Currently all addons are defined as Application. Don't want to change them right now.

**Comment by chivalryq** [comment]
> `application` should be wrapped in `output`?

we can do that to align with other definitions' usage

**Comment by hongchaodeng** [comment]
@chivalryq 

不论是 CUE 还是 YAML 都没有本质区别 -- 它们都是表达数据的一种格式，也不是解决这个问题的关键。
解决这个问题最核心的点在于我们如何去统一描述 Addon，还有如何去 Execute?
对于上面的 CUE 文件，你需要阐述为何用你的那套描述能更好地解决这两个问题？

**Comment by hongchaodeng** [comment]
@chivalryq 

Whether to use CUE to describe it is not an issue here. The biggest issue that this proposal tries to solve is how we can manage parameters, information, and installation steps of an addon. More specifically, how to implement them. Once we have defined that, it is not a big issue to describe it in CUE or YAML.

File: design/vela-core/addon.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    url: https://kubevela.io/docs/end-user/components/kustomize
+
+  dependencies:
+  - namespace: addon_namespace
+    name: addon_name
+
+  # Here defines the addon installation stepps and parameterization in CUE.
+  # This would provide flexibility and consistency for the addon management system within KubeVela.
+  template: |
+    apply: op.#Apply {
```
Comment (chivalryq) [inline design/vela-core/addon.md:67]: Style is not important. It seems we don't need a workflow-step-style template here. we can just leave an "output" and a "parameters" here. like:

```yaml
template: |-
 output: {
  metadata: {
   name:      "terraform"
   namespace: "vela-system"
  }
  spec: {
   components: {
    xxx: yyy
   }
   workflow: {
    ...
   }
  }
}
parameter: {
 appliedNamespace: *"vela-system" | string
}
```

**Comment by wonderflow** [comment]
closing as we have implemented the addon without AddonDefinition

**Pull request closed without merging**